File: pyround1/18.四数之和.py
#
# @lc app=leetcode.cn id=18 lang=python
#
# [18] 四数之和
#

# @lc code=start
class Solution(object):
    def fourSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        res = []
        nums.sort()
        n = len(nums)
        p = 0
        while p < n - 3:
            if nums[p] + 3*nums[p+1] > target: break
            if nums[p] + 3*nums[-1] < target:
                while p < n -4 and nums[p] == nums[p+1]: p += 1
                p += 1
                continue
            k = p + 1
            while k < n - 2:
                if nums[p] + nums[k] + 2 * nums[k+1] > target: break
                if nums[p] + nums[k] + 2 * nums[-1] < target:
                    while k < n - 3 and nums[k] == nums[k+1]: k += 1
                    k += 1
                    continue
                i = k + 1
                j = n - 1
                new_target = target - nums[p] - nums[k]
                while i < j:
                    if nums[i] + nums[j] > new_target: j -=1
                    elif nums[i] + nums[j] < new_target: i += 1
                    else:
                        res.append([nums[p], nums[k], nums[i], nums[j]])
                        i += 1
                        j -= 1
                        while i < j and nums[i] == nums[i-1]: i += 1
                        while i < j and nums[j] == nums[j+1]: j -= 1
                while k < n - 3 and nums[k] == nums[k+1]: k += 1
                k += 1
            while p < n -4 and nums[p] == nums[p+1]: p += 1
            p += 1
        return res
# @lc code=end

File: 18.四数之和.cpp
/*
 * @lc app=leetcode.cn id=18 lang=cpp
 *
 * [18] 四数之和
 */

// @lc code=start
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        if(nums.size() < 4)
            return {};
        for(int i = 0; i < nums.size() - 2; i++){
            if(i > 0 and nums[i] == nums[i-1])
                continue;
            int sum = target - nums[i];
            for(int j = i+1; j < nums.size() -1; j++){
                if(j > i+1 and nums[j] == nums[j-1])
                    continue;
                int left = j+1,right =nums.size() -1; 
                while(left < right){
                    if(nums[j] + nums[left] + nums[right] == sum){
                        res.push_back({nums[i], nums[j], nums[left], nums[right]});
                        while(left < right and nums[left+1] == nums[left])
                        left++;
                        while(left < right and nums[right-1] == nums[right])
                        right--;
                        left++;
                        right--; 
                    }
                    else if(nums[j] + nums[left] + nums[right]< sum){
                        left++;
                    }
                    else{
                        right--;
                    }
                }
            }
        }
        return res;

    }
};
// @lc code=end

